### PR TITLE
Remove the Garbage struct (closes #12)

### LINKED ISSUE #12: Can we think of a better name for `Garbage`?
@stjepang said:

> "Garbage" is typically in plural and refers to a bunch of deleted/dead/unlinked objects. I'd prefer to use a better synonym in singular form. That said, I don't have any ideas for a better word to use. :)

@Vtec234 replied:

> It seems that such a word literally doesn't exist: [link](https://english.stackexchange.com/questions/240876/single-countable-word-for-trash-garbage). The closest name is PieceOfGarbage.

### PR DESCRIPTION
**Pull request by jeehoonkang**
There was a discussion on the naming of `Garbage` (#12).  This PR simply removes it..

### CODE REVIEW

File: src/internal.rs
Comment (Thomasdezeeuw) [inline src/internal.rs]: Since `Deferred` isn't `Sync` or `Send` is this ok? If so a comment explaining nice would be nice.

File: src/internal.rs
Comment (Thomasdezeeuw) [inline src/internal.rs:21]: The following two paragraphs uses double spaces after a period, however the other text doesn't.

File: src/internal.rs
Comment (Thomasdezeeuw) [inline src/internal.rs]: Doc mentions `true`, but it actually returns `Ok(())`.

**Comment by jeehoonkang** [comment]
@Thomasdezeeuw Thanks!  I revised the PR as you said.

Actually, `Bag` didn't need to be `Sync`.

File: src/internal.rs
Comment (Vtec234) [inline src/internal.rs]: I can't see anywhere where that could happen, but are we absolutely sure a `Deferred` is never dropped as a standalone object rather than when it's inside a `Bag`? This change would then make it never run, as opposed to having the individual `Garbage` destructor take care of it before the change.

**Review by Vtec234 — APPROVED** [review]
Looks good, left one question.

File: src/internal.rs
Comment (Amanieu) [inline src/internal.rs]: Wouldn't it make more sense to `impl Send` for `Deferred` instead? It will then automatically be applied to `Bag`.

File: src/internal.rs
Comment (jeehoonkang) [reply src/internal.rs]: I revised the doc.  Would you please review again to check if it's clearer?

File: src/internal.rs
Comment (jeehoonkang) [reply src/internal.rs]: I'd like to cc @stjepang.  He said that [`Deferred` is better not to implement `Send`](https://github.com/crossbeam-rs/crossbeam-epoch/issues/12#issuecomment-336735548).

File: src/internal.rs
Comment (Vtec234) [reply src/internal.rs]: Thank you for revising, but I think that I had something slightly different in mind. Basically, after this change, `Deferred` relies on being inside a `Bag` to be destroyed and have its stored function run. On its own, a `Deferred` won't run when dropped. I think that's fine, but just wanted to make sure.

File: src/internal.rs
Comment (ghost) [inline src/internal.rs]: This should be `"sanitize"`. :)

File: src/internal.rs
Comment (ghost) [reply src/internal.rs]: Actually, I agree with @Amanieu. Let's have `impl Send for Deferred {}`

File: src/internal.rs
Comment (ghost) [inline src/internal.rs]: Maybe change to "A bag of deferred functions"?

File: src/internal.rs
Comment (ghost) [inline src/internal.rs]: s/garbage object/deferred function/

File: src/internal.rs
Comment (ghost) [inline src/internal.rs]: s/unsuccessful/the bag is full/

File: src/internal.rs
Comment (ghost) [inline src/internal.rs]: I don't think this actually affects safety. If you call a deferred function twice, the second call will simply panic.

File: src/internal.rs
Comment (jeehoonkang) [reply src/internal.rs]: I agree with @Vtec234, so I just changed `Deferred::call()` so that it consumes `self`.  Basically `Deferred`'s API is almost the same with `FnOnce<()>`, and so we don't need to worry about double calling.

cc @stjepang, as he is the original author of `Deferred`.  May I ask why did you allow to call `Deferred` twice (and panic)?

Comment (ghost) [reply src/internal.rs]: Maybe at the time we had to be able to do `.call()` without consuming the `Deferred`? I don't remember the details.

If you can change `call` to consume `self`, go ahead. That'd be nicer than what we currently have. 👍

**Review by ghost — APPROVED** [review]

File: src/deferred.rs
Comment (Thomasdezeeuw) [inline src/deferred.rs]: Is there a `Send`/`Sync` for functions? Does `F: FnOnce() + Send + Sync` work, or is this already the case?

Comment (Amanieu) [reply src/deferred.rs]: In this case we only need `F: FnOnce() + Send`.

Comment (Thomasdezeeuw) [reply src/deferred.rs]: @Amanieu I think you're right `Sync` isn't needed.

**Comment by jeehoonkang** [comment]
I made a significant change after this PR is approved.  Now `Deferred` is not `Send`, and the user of `Guard::defer()` should guarantee that it is safe for another thread to call the deferred function after the grace period.  I did it for giving a safe abstraction for unsafe implementations.

Let me explain a little bit more.  First of all, consider the following code snippet, which is the typical way of deferring the deallocation of the `shared` reference nowadays:

```rust
guard.defer(move || shared.into_owned());
```

Strictly speaking, we cannot require `F: Send` in `Deferred::new()`, because the function given above is actually not `Send`.  We also cannot require the user of `Deferred::new()` to guarantee that it is safe for another thread to call the function, because it's not the case for the function given above.  The (un)safety contract should be made in `Guard::defer()`, where the user should guarantee that it is safe for any thread to call the deferred function after the grace period.

So I made `Deferred` just an optimized container of `FnOnce()`, and let `Guard::defer()` to deal with thread safety.

I believe this change matches with @stjepang's original intent.  What do you think?

cc @Thomasdezeeuw

File: src/deferred.rs
Comment (Thomasdezeeuw) [inline src/deferred.rs]: Doc still mentions a panic if it was already called, but that is no longer possible.

**Review by Thomasdezeeuw — APPROVED** [review]

**Comment by ghost** [comment]
@jeehoonkang That looks good to me, I like the latest changes!

Bikeshedding time 🚲 - if you need ideas for a fun naming scheme, here's one:

```rust
impl Bag {
    fn seal(self, epoch: Epoch) -> SealedBag {
        SealedBag { epoch, bag: self }
    }
}

struct SealedBag {
    epoch: Epoch,
    bag: Bag,
}

impl SealedBag {
    fn is_expired(self, epoch: Epoch) -> bool {
        epoch.wrapping_sub(self.epoch) >= 2
    }
}

// ...

self.queue.push(bag.seal(epoch), guard);

// ...

for _ in 0..steps {
    match self.queue.try_pop_if(
        &|sealed_bag| sealed_bag.is_expired(global_epoch),
        guard,
    ) {
        // ...
    }
}
```

But the current struct/function names are ok, too.

**Comment by jeehoonkang** [comment]
@stjepang Thanks for the names.  I updated this PR as you said.

I think finally we produced a mergeable PR :)  Thanks all for your valuable comments!  I'll merge this branch once the CI becomes green.

**Pull request merged**
